Sources/PrimerSDK/Classes/PCI/User Interface/Text Fields/PrimerCardNumberFieldView.swift

```diff
@@ -33,11 +33,32 @@ public final class PrimerCardNumberFieldView: PrimerTextFieldView {
     public override func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
         guard let primerTextField = textField as? PrimerTextField else { return true }
         let currentText = primerTextField.internalText ?? ""
+
+        // Restrict to 19 digits without spaces
         if string != "" && currentText.withoutWhiteSpace.count == 19 { return false }
+
+        // Get the new text after the replacement
         let newText = (currentText as NSString).replacingCharacters(in: range, with: string) as String
+
+        // Ensure the new text is numeric (except for spaces)
         if !newText.withoutWhiteSpace.isNumeric && !string.isEmpty { return false }
+
         primerTextField.internalText = newText
 
+        // Format the text by adding spaces every 4 digits
+        let formattedText = newText.withoutWhiteSpace.separate(every: 4, with: " ")
+
+        // Set the formatted text back to the text field
+        primerTextField.text = formattedText
+
+        // Move the cursor to the last character of the text after pasting using Dispatch
+        DispatchQueue.main.async {
+            if let endPosition = primerTextField.position(from: primerTextField.endOfDocument, offset: 0) {
+                primerTextField.selectedTextRange = primerTextField.textRange(from: endPosition, to: endPosition)
```

Comment:
Won't this also move the cursor to the end if the user edits the middle of the card string? e.g. if they missed a few numbers

Reply:
true. hmmm...